onnxruntime/core/mlas/lib/fp16_common.h

```diff
@@ -0,0 +1,302 @@
+/*++
+
+Copyright (c) Microsoft Corporation.  All rights reserved.
+
+Licensed under the MIT License.
+
+Module Name:
+
+    fp16_common.h
+
+Abstract:
+
+    Intrinsic and inline functions for fp16 processing.
+
+--*/
+
+#pragma once
+
+#include "mlasi.h"
+
+#ifdef MLAS_F16VEC_INTRINSICS_SUPPORTED
+
+// TODO!! Add intel fp16 implementations
+
+typedef float16x8_t MLAS_FLOAT16X8;
+typedef float16x4_t MLAS_FLOAT16X4;
+typedef uint16x8_t MLAS_UINT16X8;
+typedef uint16x4_t MLAS_UINT16X4;
+
+#include "mlas_float16.h"
+
+MLAS_FORCEINLINE
+MLAS_FLOAT16X8
+MlasReinterpretAsFloat16x8(MLAS_INT32X4 Vector) { return vreinterpretq_f16_s32(Vector); }
+
+MLAS_FORCEINLINE
+MLAS_FLOAT16X8
+MlasBroadcastFloat16x8(_mlas_fp16_ Value) { return vreinterpretq_f16_p16(vdupq_n_p16(Value)); }
+
+MLAS_FORCEINLINE
+MLAS_FLOAT16X4
+MlasBroadcastFloat16x4(_mlas_fp16_ Value) { return vreinterpret_f16_p16(vdup_n_p16(Value)); }
+
+MLAS_FORCEINLINE
+MLAS_FLOAT16X8
+MlasBroadcastFloat16x8(const _mlas_fp16_* Value) { return vreinterpretq_f16_u16(vld1q_dup_u16(Value)); }
+
+MLAS_FORCEINLINE
+MLAS_FLOAT16X4
+MlasBroadcastFloat16x4(const _mlas_fp16_* Value) { return vreinterpret_f16_u16(vld1_dup_u16(Value)); }
+
+MLAS_FORCEINLINE
+MLAS_FLOAT16X8
+MlasZeroFloat16x8(void) { return vreinterpretq_f16_f32(vdupq_n_f32(0.0f)); }
+
+MLAS_FORCEINLINE
+MLAS_FLOAT16X8
+MlasLoadFloat16x8(const _mlas_fp16_* Buffer) { return vreinterpretq_f16_u16(vld1q_u16(Buffer)); }
+
+MLAS_FORCEINLINE
+MLAS_FLOAT16X4
+MlasLoadFloat16x4(const _mlas_fp16_* Buffer) { return vreinterpret_f16_u16(vld1_u16(Buffer)); }
+
+MLAS_FORCEINLINE
+void
+MlasStoreFloat16x8(_mlas_fp16_* Buffer, MLAS_FLOAT16X8 Vector)
+{
+    vst1q_u16(Buffer, vreinterpretq_u16_f16(Vector));
+}
+
+MLAS_FORCEINLINE
+void
+MlasStoreFloat16x4(_mlas_fp16_* Buffer, MLAS_FLOAT16X4 Vector)
+{
+    vst1_u16(Buffer, vreinterpret_u16_f16(Vector));
+}
+
+MLAS_FORCEINLINE
+void
+MlasStorePartialFloat16x4(_mlas_fp16_* Buffer, MLAS_FLOAT16X4 Vector, size_t len)
+{
+    if ((len & 2) != 0) {
+        vst1_lane_f32(reinterpret_cast<float*>(Buffer), vreinterpret_f32_f16(Vector), 0);
+        Vector = vreinterpret_f16_f32(vdup_lane_f32(vreinterpret_f32_f16(Vector), 1));
+        Buffer += 2;
+    }
+    if ((len & 1) != 0) {
+        vst1_lane_u16(Buffer, vreinterpret_u16_f16(Vector), 0);
+    }
+}
+
+template <unsigned Lane>
+MLAS_FORCEINLINE void
+MlasStoreLaneFloat16x8(_mlas_fp16_* Buffer, MLAS_FLOAT16X8 Vector)
+{
+    vst1q_lane_u16(Buffer, vreinterpretq_u16_f16(Vector), Lane);
+}
+
+MLAS_FORCEINLINE MLAS_FLOAT16X4
+MlasToLowHalfFloat16x4(MLAS_FLOAT16X8 V)
+{
+    // vget_low should be compiled to nothing
+    // https://godbolt.org/z/ad2fVM
```

Comment:
Is this a permanent link to a permanent page? Should we just paste what generated by compiler instead of a external link here?